src/memo/memo.ml

```diff
@@ -1344,18 +1344,59 @@ let get_call_stack = Call_stack.get_call_stack_without_state
 module Invalidation = struct
   type ('i, 'o) memo = ('i, 'o) t
 
-  module Leaf = struct
+  (* This is currently used only for informing the user about the reason for
+     restarting a build. *)
+  module Reason = struct
+    (* CR-someday amokhov: Add other reasons, e.g. a watched environment
+       variable changed. *)
     type t =
-      | Invalidate_node : _ Dep_node.t -> t
-      | Clear_cache : ('input, ('input, 'output) Dep_node.t) Store.t -> t
+      | Unknown
+      | Path_changed of Path.t
+      | Event_queue_overflow
+      | Upgrade
+      | Test
+
+    let to_dyn = function
+      | Unknown -> Dyn.Variant ("Unknown", [])
+      | Path_changed path -> Dyn.Variant ("Path_changed", [ Path.to_dyn path ])
+      | Event_queue_overflow -> Dyn.Variant ("Event_queue_overflow", [])
+      | Upgrade -> Dyn.Variant ("Upgrade", [])
+      | Test -> Dyn.Variant ("Test", [])
+
+    let to_string_hum = function
+      | Unknown -> None
```

Comment:
I'd print "Unknown reason" here instead of nothing, to avoid having to remember that no reason printed means an unknown reason.

Reply:
At least when there are no other reasons.